core/staking_verifier.go

```diff
@@ -0,0 +1,265 @@
+package core
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+
+	"github.com/pkg/errors"
+
+	"github.com/harmony-one/harmony/core/vm"
+	common2 "github.com/harmony-one/harmony/internal/common"
+	"github.com/harmony-one/harmony/numeric"
+	"github.com/harmony-one/harmony/shard"
+	staking "github.com/harmony-one/harmony/staking/types"
+)
+
+// TODO: add unit tests to check staking msg verification
+
+// VerifyCreateValidatorStakingMsg verifies the create validator message using
+// the stateDB and epoch if they are provided.
+// It is not necessary to provide the stateDB and/or epoch; validation that
+// does not using said components will still run.
+//
+// Note that these checks never update the stateDB, it only reads from stateDB.
+func VerifyCreateValidatorStakingMsg(
+	stateDB vm.StateDB, epoch *big.Int, msg *staking.CreateValidator,
+) error {
+	if msg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+	if stateDB != nil {
+		if stateDB.IsValidator(msg.ValidatorAddress) {
+			return errors.Wrapf(errValidatorExist, common2.MustAddressToBech32(msg.ValidatorAddress))
+		}
+		if !CanTransfer(stateDB, msg.ValidatorAddress, msg.Amount) {
+			return errInsufficientBalanceForStake
+		}
+	}
+	// create temp validator to do sanity check
+	v, err := staking.CreateValidatorFromNewMsg(msg, new(big.Int).SetInt64(0))
+	if err != nil {
+		return err
+	}
+	zero := big.NewInt(0)
+	wrapper := staking.ValidatorWrapper{}
+	wrapper.Validator = *v
+	wrapper.Delegations = []staking.Delegation{
+		staking.NewDelegation(v.Address, msg.Amount),
+	}
+	wrapper.Snapshot.Epoch = epoch
+	wrapper.Snapshot.NumBlocksSigned = zero
+	wrapper.Snapshot.NumBlocksToSign = zero
+	return wrapper.SanityCheck()
+}
+
+// VerifyEditValidatorStakingMsg verifies the edit validator message using
+// the stateDB and chainContext if they are provided.
+// It is not necessary to provide the stateDB and/or chainContext; validation that
+// does not using said components will still run.
```

Comment:
"does not use". Please correct everywhere.

Reply:
corrected.